x/cronos/keeper/evm.go

```diff
@@ -0,0 +1,154 @@
+
+	"github.com/crypto-org-chain/cronos/x/cronos/types"
+)
+
+// CallEVM execute an evm message from native module
+func (k Keeper) CallEVM(ctx sdk.Context, to *common.Address, data []byte, value *big.Int) (*ethtypes.Message, *evmtypes.MsgEthereumTxResponse, error) {
+	k.evmKeeper.WithContext(ctx)
+
+	nonce := k.evmKeeper.GetNonce(types.EVMModuleAddress)
+	msg := ethtypes.NewMessage(
+		types.EVMModuleAddress,
+		to,
+		nonce,
+		value,         // amount
+		25000000,      // gasLimit, FIXME configuration
+		big.NewInt(0), // gasPrice
+		data,
+		nil,   // accessList
+		false, // checkNonce
+	)
+
+	params := k.evmKeeper.GetParams(ctx)
+	// return error if contract creation or call are disabled through governance
+	if !params.EnableCreate && to == nil {
+		return nil, nil, errors.New("failed to create new contract")
+	} else if !params.EnableCall && to != nil {
+		return nil, nil, errors.New("failed to call contract")
+	}
+	ethCfg := params.ChainConfig.EthereumConfig(k.evmKeeper.ChainID())
+
+	// get the coinbase address from the block proposer
+	coinbase, err := k.evmKeeper.GetCoinbaseAddress(ctx)
+	if err != nil {
+		return nil, nil, errors.New("failed to obtain coinbase address")
+	}
+	evm := k.evmKeeper.NewEVM(msg, ethCfg, params, coinbase, nil)
+	ret, err := k.evmKeeper.ApplyMessage(evm, msg, ethCfg, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	k.evmKeeper.CommitCachedContexts()
+	return &msg, ret, nil
+}
+
+// CallCronosERC20 call a method of CronosERC20 contract
+func (k Keeper) CallCronosERC20(ctx sdk.Context, contract common.Address, method string, args ...interface{}) ([]byte, error) {
+	data, err := types.CronosERC20Contract.ABI.Pack(method, args...)
+	if err != nil {
+		return nil, err
+	}
+	_, res, err := k.CallEVM(ctx, &contract, data, big.NewInt(0))
+	if err != nil {
+		return nil, err
+	}
+	if res.Failed() {
+		return nil, fmt.Errorf("call contract failed: %s, %s, %s", contract.Hex(), method, res.Ret)
+	}
+	return res.Ret, nil
+}
+
+// DeployCronosERC20 deploy an embed erc20 contract
+func (k Keeper) DeployCronosERC20(ctx sdk.Context, denom string) (common.Address, error) {
+	ctor, err := types.CronosERC20Contract.ABI.Pack("", denom, uint8(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+	data := append(types.CronosERC20Contract.Bin, ctor...)
+
+	msg, res, err := k.CallEVM(ctx, nil, data, big.NewInt(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if res.Failed() {
+		return common.Address{}, fmt.Errorf("contract deploy failed: %s", res.Ret)
+	}
+	return crypto.CreateAddress(types.EVMModuleAddress, msg.Nonce()), nil
+}
+
+// SendCoinFromNativeToERC20 convert native token to erc20 token
+func (k Keeper) SendCoinFromNativeToERC20(ctx sdk.Context, sender common.Address, coin sdk.Coin) error {
+	// TODO validate coin.Denom
+	var err error
+	contract, found := k.GetContractByDenom(ctx, coin.Denom)
+	if !found {
+		contract, err = k.DeployCronosERC20(ctx, coin.Denom)
```

Comment:
hmm I dont know if we should support this feature

Reply:
if that's the case, it defeats the purpose of most parts of this PR. 😂

Reply:
mm maybe you can just add it for the testing framework?

Reply:
Since we are not sure in the short term, but could need it in the longer term, I added a bool parameter to this method, make the auto-deploy part optional.